eip-0030.md

```diff
@@ -0,0 +1,329 @@
+    //   R1 (value): X tokens in NanoErgs 
+    //   R4: How many LP in circulation (long). This can be non-zero when bootstrapping, to consider the initial token burning in UniSwap v2
+    //   R5: Cross-counter. A counter to track how many times the rate has "crossed" the oracle pool rate. That is the oracle pool rate falls in between the before and after rates
+    //   Tokens(0): LP NFT to uniquely identify NFT box. (Could we possibly do away with this?) 
+    //   Tokens(1): LP tokens
+    //   Tokens(2): Y tokens (Note that X tokens are NanoErgs (the value) 
+    //   
+    // Data Input #0: (oracle pool box)
+    //   R4: Rate in units of X per unit of Y
+    //   Token(0): OP NFT to uniquely identify Oracle Pool
+     
+    // constants 
+    val feeNum = 3 // 0.3 % 
+    val feeDenom = 1000
+    val minStorageRent = 10000000L  // this many number of nanoErgs are going to be permanently locked
+    
+    val successor = OUTPUTS(0) // copy of this box after exchange
+    val oraclePoolBox = CONTEXT.dataInputs(0) // oracle pool box
+    val validOraclePoolBox = oraclePoolBox.tokens(0)._1 == fromBase64("RytLYlBlU2hWbVlxM3Q2dzl6JEMmRilKQE1jUWZUalc=") // to identify oracle pool box 
+    
+    val lpNFT0    = SELF.tokens(0)
+    val reservedLP0 = SELF.tokens(1)
+    val tokenY0     = SELF.tokens(2)
+
+    val lpNFT1    = successor.tokens(0)
+    val reservedLP1 = successor.tokens(1)
+    val tokenY1     = successor.tokens(2)
+
+    val supplyLP0 = SELF.R4[Long].get       // LP tokens in circulation in input LP box
+    val supplyLP1 = successor.R4[Long].get  // LP tokens in circulation in output LP box
+
+    val validSuccessorScript = successor.propositionBytes == SELF.propositionBytes
+    
+    val preservedLpNFT     = lpNFT1 == lpNFT0
+    val validLP              = reservedLP1._1 == reservedLP0._1
+    val validY               = tokenY1._1 == tokenY0._1
+    val validSupplyLP1       = supplyLP1 >= 0
+       
+    // since tokens can be repeated, we ensure for sanity that there are no more tokens
+    val noMoreTokens         = successor.tokens.size == 3
+  
+    val validStorageRent     = successor.value > minStorageRent
+
+    val reservesX0 = SELF.value
+    val reservesY0 = tokenY0._2
+    val reservesX1 = successor.value
+    val reservesY1 = tokenY1._2
+
+    val oraclePoolRateXY = oraclePoolBox.R4[Long].get 
+    val lpRateXY0 = reservesX0 / reservesY0  // we can assume that reservesY0 > 0 (since at least one token must exist) 
+    val lpRateXY1 = reservesX1 / reservesY1  // we can assume that reservesY1 > 0 (since at least one token must exist)
+    val isCrossing = (lpRateXY0 - oraclePoolRateXY) * (lpRateXY1 - oraclePoolRateXY) < 0 // if (and only if) oracle pool rate falls in between, then this will be negative
+     
+    val crossCounterIn = SELF.R5[Int].get
+    val crossCounterOut = successor.R5[Int].get
+    
+    val validCrossCounter = crossCounterOut == {if (isCrossing) crossCounterIn + 1 else crossCounterIn}
```

Comment:
Why to check this in LP contract?

Reply:
Just to keep track of how many times its crossed (see above comment). Note that the requirement we are trying to capture is that between a top-up initiation and top-up completion, the LP rate should not have crossed the oracle pool rate. If it does, then the top-up becomes invalid and we need to start afresh.